proposals/csharp-10.0/parameterless-struct-constructors.md

```diff
@@ -0,0 +1,106 @@
+# Parameterless struct constructors
+
+## Summary
+
+Support explicit parameterless constructors for value types.
+
+## Proposal
+
+If the parameterless constructor is less accessible than the value type, a warning is reported and the constructor is ignored when inaccessible.
```

Comment:
Does such a `struct` satisfy the `new()` constraint?

Reply:
I did some testing of this, and I don't believe that it should. `Activator.CreateInstance` does not work if the struct constructor is declared as `internal` or `private`, regardless of whether the containing type is `internal` or not. It will _only_ call `public` constructors. We enforce this for reference types today:

```cs
M<C>(); // error CS0310: 'C' must be a non-abstract type with a public parameterless constructor in order to use it as parameter 'T' in the generic type or method 'M<T>()'

void M<T>() where T : new()
{
    _ = new T();
}

internal class C {
    internal C() { }
}
```

Reply:
Yes, thanks for catching this.

---
In reply to: [581598691](https://github.com/dotnet/csharplang/pull/4459#discussion_r581598691) [](ancestors = 581598691)